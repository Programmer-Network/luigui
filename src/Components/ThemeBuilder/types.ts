export interface IColors {
  primary: string;
  background: string;
  text: string;
}

export interface IThemeBuilder {
  onColorsChange: (colors: IColors) => void;
  onReset: () => void;
  isResetButtonShown: boolean;
  settings: {
    primary: string;
    background: string;
    text: string;
  };
  buttons: {
    label: string;
    type: string;
    cssVariable: string;
    className?: string;
  }[];
}
