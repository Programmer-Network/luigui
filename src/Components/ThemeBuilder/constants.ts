export const initialSelected = {
  type: "",
  color: "",
  cssVariable: ""
};
