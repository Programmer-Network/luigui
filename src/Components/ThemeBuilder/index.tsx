import { useOnClickOutside } from "Hooks/useClickOutside";
import classNames from "classnames";
import { FC, useEffect, useMemo, useRef, useState } from "react";
import { HexColorInput, HexColorPicker } from "react-colorful";

import Button from "Components/Button";
import Icon from "Components/Icon";

import { IColors, IThemeBuilder } from "./types";

const ThemeBuilder: FC<IThemeBuilder> = ({
  onColorsChange,
  onReset,
  isResetButtonShown,
  buttons,
  settings
}) => {
  const initialSelected = {
    type: "",
    color: "",
    cssVariable: ""
  };

  const [selected, setSelected] = useState<{
    type: string;
    color: string;
    cssVariable: string;
  }>(initialSelected);

  const [colors, setColors] = useState<IColors>({
    primary: "",
    background: "",
    text: ""
  });

  const initialSelectedStateRef = useRef(settings);
  const ref = useRef<HTMLDivElement>(null);

  const isDirty = useMemo(
    () =>
      Object.keys(colors).some(
        key =>
          colors[key as keyof IColors] !==
          initialSelectedStateRef.current[key as keyof IColors]
      ),
    [colors]
  );

  useOnClickOutside(ref, () => {
    setSelected(initialSelected);
  });

  const handleSaveColors = () => {
    onColorsChange(colors);
    setColors(colors);
    setSelected(initialSelected);
  };

  const setCSSVariable = (variable: string, value: string) =>
    document.documentElement.style.setProperty(variable, value);

  const handleReset = () => {
    setCSSVariable("--color-primary", "#4f46e5");
    setCSSVariable("--text-color", "#6b7280");
    setCSSVariable("--color-bg", "#1b1f23");
    setSelected(initialSelected);
    setColors(settings);
    onReset();
  };

  useEffect(() => {
    setCSSVariable("--color-primary", settings.primary);
    setCSSVariable("--text-color", settings.text);
    setCSSVariable("--color-bg", settings.background);
    setColors(settings);
  }, [settings.primary, settings.background, settings.text]);

  useEffect(() => {
    if (!selected.type) {
      return;
    }

    setCSSVariable(selected.cssVariable, selected.color);
    setColors({
      ...colors,
      [selected.type]: selected.color
    });
  }, [selected.type, selected.color]);

  return (
    <div ref={ref}>
      <div className='mb-1 flex items-center'>
        <div
          className={classNames("border-2 border-r-0 border-primary px-3 py-2")}
        >
          <Icon iconName='IconColors' className='w-5 text-primary' />
        </div>
        {buttons.map(button => {
          return (
            <Button
              key={button.type}
              outlined={button.type !== selected.type}
              className={button.className}
              onClick={() => {
                setSelected({
                  type: button.type,
                  cssVariable: button.cssVariable,
                  color: colors[button.type as keyof typeof colors]
                });
              }}
            >
              {button.label}
            </Button>
          );
        })}
      </div>
      {isDirty && (
        <div className='flex gap-1'>
          <Button className='group' onClick={handleSaveColors}>
            Save
          </Button>

          {isResetButtonShown && (
            <Button outlined onClick={handleReset}>
              Reset
            </Button>
          )}
        </div>
      )}
      {selected.type && (
        <div className='relative flex items-center justify-start'>
          <div className='relative mt-4'>
            <div onClick={() => setSelected(initialSelected)}>
              <Icon
                iconName='IconClose'
                className='absolute -right-4 -top-4 z-50 w-4 cursor-pointer text-primary'
              />
            </div>

            <HexColorPicker
              onChange={color => setSelected({ ...selected, color })}
              color={selected?.color}
            />
            <HexColorInput
              role='presentation'
              onChange={(color: string) => setSelected({ ...selected, color })}
              color={selected?.color || ""}
              placeholder='e.g.175ca1'
              className='bg mt-2 rounded-lg border-2 border-primary bg-transparent p-2 text-primary-text-color shadow-md hover:border-primary focus:border-primary focus:shadow-none focus:outline-none focus:ring-transparent dark:text-primary-text-color'
            />
          </div>
        </div>
      )}
    </div>
  );
};

export default ThemeBuilder;
