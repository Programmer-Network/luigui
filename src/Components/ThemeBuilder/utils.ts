export const setCSSVariable = (variable: string, value: string) =>
  document.documentElement.style.setProperty(variable, value);
