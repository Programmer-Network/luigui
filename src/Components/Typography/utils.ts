export const headingCommonStyles = "font-bold text-primary mb-4";
