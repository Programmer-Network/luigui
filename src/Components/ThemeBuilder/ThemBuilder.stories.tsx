import { useState } from "react";

import { H1, Paragraph } from "Components/Typography";

import ThemeBuilder from ".";

export default {
  title: "Components / Theme Builder"
};

export const Primary = () => {
  const [settings, setSettings] = useState({
    primary: "#4f46e5",
    background: "#1b1f23",
    text: "#6b7280"
  });

  return (
    <div className='m-10'>
      <ThemeBuilder
        isResetButtonShown={true}
        onColorsChange={setSettings}
        onReset={() => null}
        settings={settings}
        buttons={[
          {
            label: "Primary",
            type: "primary",
            cssVariable: "--color-primary"
          },
          {
            label: "Background",
            type: "background",
            cssVariable: "--color-bg",
            className: "border-l-0 border-r-0"
          },
          {
            label: "Text",
            type: "text",
            cssVariable: "--text-color"
          }
        ]}
      />

      <section className='bg-primary-background-color p-6'>
        <H1>Lorem ipsum dolor sit amet.</H1>
        <Paragraph>
          Lorem ipsum dolor, sit amet consectetur adipisicing elit. Harum beatae
          veritatis exercitationem dolores. Eos saepe tenetur, adipisci deserunt
          quasi sit aliquam explicabo voluptas id nobis omnis quaerat aspernatur
          perspiciatis provident itaque ea eveniet fuga vero! Maiores quas rem
          voluptatibus consectetur ipsum! Asperiores suscipit reprehenderit
          porro tempore corrupti iure deserunt distinctio.
        </Paragraph>
      </section>
    </div>
  );
};
