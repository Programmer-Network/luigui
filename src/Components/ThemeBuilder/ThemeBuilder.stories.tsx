import { useState } from "react";

import { H1, Paragraph } from "Components/Typography";

import ThemeBuilder from ".";

export default {
  title: "Components / Theme Builder"
};

export const Default = () => {
  const [settings, setSettings] = useState([
    {
      label: "Primary",
      type: "primary",
      customCSSProperty: "--color-primary",
      color: "79 70 229",
      defaultColor: "79 70 229"
    },
    {
      label: "Background",
      type: "background",
      customCSSProperty: "--color-bg",
      color: "27 31 35",
      defaultColor: "27 31 35"
    },
    {
      label: "Text",
      type: "text",
      customCSSProperty: "--text-color",
      color: "107 114 128",
      defaultColor: "107 114 128"
    }
  ]);

  return (
    <div className='m-12'>
      <ThemeBuilder
        onChange={setSettings}
        onReset={() => null}
        settings={settings}
      />

      <section className='bg-primary-background-color'>
        <H1>Lorem ipsum dolor sit amet.</H1>
        <Paragraph>
          Lorem ipsum dolor, sit amet consectetur adipisicing elit. Harum beatae
          veritatis exercitationem dolores. Eos saepe tenetur, adipisci deserunt
          quasi sit aliquam explicabo voluptas id nobis omnis quaerat aspernatur
          perspiciatis provident itaque ea eveniet fuga vero! Maiores quas rem
          voluptatibus consectetur ipsum! Asperiores suscipit reprehenderit
          porro tempore corrupti iure deserunt distinctio.
        </Paragraph>
      </section>
    </div>
  );
};
