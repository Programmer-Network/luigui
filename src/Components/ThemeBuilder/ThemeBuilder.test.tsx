import { fireEvent, render, screen } from "@testing-library/react";
import { vi } from "vitest";

import ThemeBuilder from ".";
import { hexToRGB } from "./utils";

describe("ThemeBuilder component", () => {
  const mockSettings = [
    {
      label: "Primary Color",
      type: "primary",
      customCSSProperty: "--color-primary",
      color: "#4f46e5",
      defaultColor: "#4f46e5",
      className: "primary-button"
    },
    {
      label: "Background",
      type: "background",
      customCSSProperty: "--color-bg",
      color: "#1b1f23",
      defaultColor: "#1b1f23"
    },
    {
      label: "Text Color",
      type: "text",
      customCSSProperty: "--text-color",
      color: "#6b7280",
      defaultColor: "#6b7280"
    }
  ];

  const mockOnColorsChange = vi.fn();
  const mockOnReset = vi.fn();

  it("renders correctly with given props", () => {
    const { asFragment } = render(
      <ThemeBuilder
        settings={mockSettings}
        onChange={mockOnColorsChange}
        onReset={mockOnReset}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders save and reset buttons when state is dirty", () => {
    render(
      <ThemeBuilder
        settings={mockSettings}
        onChange={mockOnColorsChange}
        onReset={mockOnReset}
      />
    );

    const primaryButton = screen.getByText("Primary Color");
    fireEvent.click(primaryButton);

    const colorPicker = screen.getByRole("presentation");
    fireEvent.change(colorPicker, { target: { value: "#ff0000" } });

    expect(screen.getByText("Save")).toBeInTheDocument();
    expect(screen.getByText("Reset")).toBeInTheDocument();
  });

  it("updates selected state on button click", () => {
    render(
      <ThemeBuilder
        settings={mockSettings}
        onChange={mockOnColorsChange}
        onReset={mockOnReset}
      />
    );

    const primaryButton = screen.getByText("Primary Color");

    fireEvent.click(primaryButton);
    expect(primaryButton.parentNode?.parentNode).toHaveClass("primary-button");
  });

  it("calls handleSaveColors on save button click", () => {
    render(
      <ThemeBuilder
        settings={mockSettings}
        onChange={mockOnColorsChange}
        onReset={mockOnReset}
      />
    );

    const primaryButton = screen.getByText("Primary Color");
    fireEvent.click(primaryButton);

    const colorPicker = screen.getByRole("presentation");
    fireEvent.change(colorPicker, { target: { value: "#ff0000" } });

    const saveButton = screen.getByText("Save");
    fireEvent.click(saveButton);
    expect(mockOnColorsChange).toHaveBeenCalled();
  });

  it("resets colors and calls onReset on reset button click", () => {
    render(
      <ThemeBuilder
        settings={mockSettings}
        onChange={mockOnColorsChange}
        onReset={mockOnReset}
      />
    );

    const primaryButton = screen.getByText("Primary Color");
    fireEvent.click(primaryButton);

    const colorPicker = screen.getByRole("presentation");
    fireEvent.change(colorPicker, { target: { value: "#ff0000" } });

    const resetButton = screen.getByText("Reset");
    fireEvent.click(resetButton);
    expect(mockOnReset).toHaveBeenCalled();
  });

  it("sets initial CSS variables based on provided settings", () => {
    render(
      <ThemeBuilder
        settings={mockSettings}
        onChange={mockOnColorsChange}
        onReset={mockOnReset}
      />
    );

    expect(
      document.documentElement.style.getPropertyValue("--color-primary")
    ).toBe(mockSettings[0].color);

    expect(document.documentElement.style.getPropertyValue("--color-bg")).toBe(
      mockSettings[1].color
    );

    expect(
      document.documentElement.style.getPropertyValue("--text-color")
    ).toBe(mockSettings[2].color);
  });

  it("updates CSS variables dynamically when a color is changed", () => {
    render(
      <ThemeBuilder
        settings={mockSettings}
        onChange={mockOnColorsChange}
        onReset={mockOnReset}
      />
    );

    const textButton = screen.getByText("Text Color");
    fireEvent.click(textButton);

    const colorPicker = screen.getByRole("presentation");
    fireEvent.change(colorPicker, { target: { value: "#abcdef" } });

    expect(
      document.documentElement.style.getPropertyValue("--text-color")
    ).toBe("#abcdef");
  });

  it("updates state and CSS variable on HexColorInput change", () => {
    render(
      <ThemeBuilder
        settings={mockSettings}
        onChange={mockOnColorsChange}
        onReset={mockOnReset}
      />
    );

    const textButton = screen.getByText("Text Color");
    fireEvent.click(textButton);

    const colorInput = screen.getByPlaceholderText("e.g.175ca1");
    fireEvent.change(colorInput, { target: { value: "#fedcba" } });

    expect(
      document.documentElement.style.getPropertyValue("--text-color")
    ).toBe("#fedcba");
  });
});

describe("hexToRGB", () => {
  it("converts six-character hex codes to RGB format", () => {
    expect(hexToRGB("#4f46e5")).toBe("79 70 229");
  });

  it("converts three-character hex codes to RGB format", () => {
    expect(hexToRGB("#4e2")).toBe("68 238 34");
  });

  it("returns null for invalid hex codes", () => {
    expect(hexToRGB("123456")).toBeNull();
    expect(hexToRGB("#gggggg")).toBeNull();
    expect(hexToRGB("#1234")).toBeNull();
  });

  it("is case-insensitive", () => {
    expect(hexToRGB("#4F46E5")).toBe("79 70 229");
    expect(hexToRGB("#4e2")).toBe("68 238 34");
  });

  it("handles edge cases", () => {
    expect(hexToRGB("#000000")).toBe("0 0 0");
    expect(hexToRGB("#ffffff")).toBe("255 255 255");
    expect(hexToRGB("#FFF")).toBe("255 255 255");
  });
});
