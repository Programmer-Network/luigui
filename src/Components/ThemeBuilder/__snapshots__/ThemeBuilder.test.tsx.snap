// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ThemeBuilder component > renders correctly with given props 1`] = `
<DocumentFragment>
  <div>
    <div
      class="mb-1 flex h-12 items-center"
    >
      <div
        class="relative flex h-full items-center justify-center bg-primary px-3 py-2"
      >
        <div>
          Loading...
        </div>
      </div>
      <button
        class="select-none border-2 border-primary px-3 py-2 font-semibold uppercase tracking-tight rounded-md bg-transparent text-primary fill-primary hover:bg-primary hover:fill-primary-background-color hover:text-primary-background-color primary-button h-12 min-w-[120px] border-l-0 shadow-none border-r-0"
        type="button"
      >
        <div
          class="relative flex items-center justify-center"
        >
          <span
            class=""
          >
            Primary Color
          </span>
          <span
            class="absolute"
          />
        </div>
      </button>
      <button
        class="select-none border-2 border-primary px-3 py-2 font-semibold uppercase tracking-tight rounded-md bg-transparent text-primary fill-primary hover:bg-primary hover:fill-primary-background-color hover:text-primary-background-color h-12 min-w-[120px] border-l-0 shadow-none border-r-0"
        type="button"
      >
        <div
          class="relative flex items-center justify-center"
        >
          <span
            class=""
          >
            Background
          </span>
          <span
            class="absolute"
          />
        </div>
      </button>
      <button
        class="select-none border-2 border-primary px-3 py-2 font-semibold uppercase tracking-tight rounded-md bg-transparent text-primary fill-primary hover:bg-primary hover:fill-primary-background-color hover:text-primary-background-color h-12 min-w-[120px] border-l-0 shadow-none"
        type="button"
      >
        <div
          class="relative flex items-center justify-center"
        >
          <span
            class=""
          >
            Text Color
          </span>
          <span
            class="absolute"
          />
        </div>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
